File: java/src/main/java/br/com/emendes/reverseinteger/ReverseInteger.java
package br.com.emendes.reverseinteger;

/**
 * Dado um inteiro x de 32 bits com sinal, retorne x com seus dígitos invertidos.
 * Se a reversão de x fizer com que o valor saia do intervalo de inteiros de 32 bits assinados [-2³¹, 2³¹ - 1], retorne 0.
 * <br><br>
 * Suponha que o ambiente não permita que você armazene números inteiros de 64 bits (com ou sem sinal).
 */
public class ReverseInteger {

  // Solução que quebra a regra de não usar inteiros 64 bits.
  public int reverse(int x) {
    boolean isNegative = x < 0;

    long reverseX = 0;
    int digit;
    if (isNegative) x *= -1;

    while (x > 0) {
      digit = x % 10;
      reverseX = (reverseX * 10) + digit;
      x = x/10;
    }

    if(isNegative) reverseX *= -1;

    if (reverseX > 2_147_483_647 || reverseX < -2_147_483_648) return 0;

    return (int) reverseX;
  }

}

File: javascript/problems/ReverseInteger.js
/**
 * Dado um inteiro x de 32-bits com sinal, retorne x com os dígitos invertidos. Se inverter x causar um valor
 * fora do intervalo de inteiro de 32-bits com sinal, então retorne 0.
 *
 * Assuma que o ambiente não permita o armazenamento de inteiros 64-bits com sinal ou sem.
 *
 * Restrições:
 *
 * -2³¹ <= x <= 2³¹ - 1
 *
 * @param {number} x
 * @return {number}
 */
var reverse = function (x) {
  const MIN = -214748364;
  const MAX = 214748364;
  const MIN_LAST_DIGIT = -8;
  const MAX_LAST_DIGIT = 7;
  const round = x < 0 ? Math.ceil : Math.floor;
  let reversedX = 0;
  while (x) {
    let digit = x % 10;
    x = round(x / 10);
    if (reversedX > MAX || (reversedX === MAX && digit > MAX_LAST_DIGIT))
      return 0;
    if (reversedX < MIN || (reversedX === MIN && digit < MIN_LAST_DIGIT))
      return 0;

    reversedX = reversedX * 10 + digit;
  }

  return reversedX;
};
